packages/flutter/lib/src/widgets/undo_history.dart

```diff
@@ -0,0 +1,446 @@
+  /// will not be called. For example, receiving a redo when there is nothing
+  /// to redo will not call this method.
+  final void Function(T value) onTriggered;
+
+  /// The [FocusNode] that will be used to listen for focus to set the initial
+  /// undo state for the element.
+  final FocusNode focusNode;
+
+  /// {@template flutter.widgets.undoHistory.controller}
+  /// Controls the undo state.
+  /// {@endtemplate}
+  final UndoHistoryController controller;
+
+  /// The child widget of [UndoHistory].
+  final Widget child;
+
+  @override
+  State<UndoHistory<T>> createState() => _UndoHistoryState<T>();
+}
+
+/// State for a [UndoHistory].
+///
+/// Provides [undo], [redo], [canUndo], and [canRedo] for programmatic access
+/// to the undo state for custom undo and redo UI implementations.
+class _UndoHistoryState<T> extends State<UndoHistory<T>> implements UndoManagerClient {
+  final _UndoStack<T> _stack = _UndoStack<T>();
+  late final _Throttled<T> _throttledPush;
+  Timer? _throttleTimer;
+
+  // This duration was chosen as a best fit for the behavior of Mac, Linux,
+  // and Windows undo/redo state save durations, but it is not perfect for any
+  // of them.
+  static const Duration _kThrottleDuration = Duration(milliseconds: 500);
+
+  // Record the last value to prevent pushing multiple
+  // of the same value in a row onto the undo stack. For example, _push gets
+  // called both in initState and when the EditableText receives focus.
+  T? _lastValue;
+
+  @override
+  void undo() {
+    _update(_stack.undo());
+    _updateState();
+  }
+
+  @override
+  void redo() {
+    _update(_stack.redo());
+    _updateState();
+  }
+
+  @override
+  bool get canUndo => _stack.canUndo;
+
+  @override
+  bool get canRedo => _stack.canRedo;
+
+  void _updateState() {
+    widget.controller.value = UndoHistoryValue(canUndo: canUndo, canRedo: canRedo);
+
+    if (defaultTargetPlatform != TargetPlatform.iOS) {
+      return;
+    }
+
+    if (UndoManager.client == this) {
+      UndoManager.setUndoState(canUndo: canUndo, canRedo: canRedo);
+    }
+  }
+
+  void _undoFromIntent(UndoTextIntent intent) {
+    undo();
+  }
+
+  void _redoFromIntent(RedoTextIntent intent) {
+    redo();
+  }
+
+  void _update(T? nextValue) {
+    if (nextValue == null) {
+      return;
+    }
+    if (nextValue == _lastValue) {
+      return;
+    }
+    _lastValue = nextValue;
+    widget.onTriggered(nextValue);
+  }
+
+  void _push() {
+    if (widget.value.value == _lastValue) {
+      return;
+    }
+
+    if (!(widget.shouldChangeUndoStack?.call(_lastValue, widget.value.value) ?? true)) {
+      return;
+    }
+
+    _lastValue = widget.value.value;
+
+    _throttleTimer = _throttledPush(widget.value.value);
+  }
+
+  void _handleFocus() {
+    if (!widget.focusNode.hasFocus) {
+      return;
+    }
+    UndoManager.client = this;
+    _updateState();
+  }
+
+  @override
+  void handlePlatformUndo(String direction) {
```

Comment:
In general I've been thinking about moving away from the pattern of passing raw, privately defined strings around from these platform methods.

Is there a way you could convert this direction to an enum or something that is publicly defined?  Then in your tests you could also reference that instead of hardcoded "undo" and "redo" strings that I saw in a few places.